pkg/rbd/rbd_journal.go

```diff
@@ -123,40 +122,55 @@ func checkSnapExists(ctx context.Context, rbdSnap *rbdSnapshot, cr *util.Credent
 		return false, nil
 	}
 	rbdSnap.RbdSnapName = snapJournal.NamingPrefix() + snapUUID
-
+	// update rbdImageName as snapshot will be created for cloned image with
+	// name RbdSnapName
+	rbdSnap.RbdImageName = rbdSnap.RbdSnapName
 	// Fetch on-disk image attributes
-	err = updateSnapWithImageInfo(ctx, rbdSnap, cr)
+
+	vol := generateVolFromSnap(rbdSnap)
+	// Fetch on-disk image attributes and compare against request
+	err = updateVolWithImageInfo(ctx, vol, cr)
 	if err != nil {
-		if _, ok := err.(ErrSnapNotFound); ok {
+		if _, ok := err.(ErrImageNotFound); ok {
 			err = snapJournal.UndoReservation(ctx, rbdSnap.Monitors, cr, rbdSnap.Pool,
 				rbdSnap.RbdSnapName, rbdSnap.RequestName)
 			return false, err
 		}
 		return false, err
 	}
 
+	ok, err := checkIsFlattenInProgress(ctx, vol, cr)
```

Comment:
This seems like you should just be calling `flattenRbdImage`. If the image is at the soft-snap limit it will just (re)schedule the background task to get it flattened. If it's at the hard-limit, we should `ABORT` while the flatten is running in the background (or block only if you cannot schedule the background flatten).